api/graphql/mutations/cohort.py

```diff
@@ -0,0 +1,171 @@
+# pylint: disable=redefined-builtin, import-outside-toplevel, ungrouped-imports
+
+from typing import TYPE_CHECKING, Annotated
+import strawberry
+from strawberry.types import Info
+
+from db.python.connect import Connection
+from db.python.filters.generic import GenericFilter
+from db.python.layers.cohort import CohortLayer
+from db.python.tables.cohort import CohortFilter
+from models.models.cohort import CohortCriteria, CohortTemplate
+from models.models.project import ProjectId, ProjectMemberRole, ReadAccessRoles
+from models.utils.cohort_template_id_format import (
+    cohort_template_id_transform_to_raw,
+)
+
+if TYPE_CHECKING:
+    from api.graphql.schema import GraphQLCohort, GraphQLCohortTemplate
+
+
+@strawberry.input
+class CohortCriteriaInput:
+    """Cohort criteria"""
+
+    projects: list[str] | None = None
+    sg_ids_internal: list[str] | None = None
+    excluded_sgs_internal: list[str] | None = None
+    sg_technology: list[str] | None = None
+    sg_platform: list[str] | None = None
+    sg_type: list[str] | None = None
+    sample_type: list[str] | None = None
+
+
+@strawberry.input
+class CohortBodyInput:
+    """Cohort body"""
+
+    name: str
+    description: str
+    template_id: str | None = None
+
+
+@strawberry.input
+class CohortTemplateInput:
+    """Cohort template input"""
+
+    id: int | None = None
+    name: str
+    description: str
+    criteria: CohortCriteriaInput
+
+
+@strawberry.type
+class CohortMutations:
+    """Cohort mutations"""
+
+    @strawberry.mutation
+    async def create_cohort_from_criteria(
+        self,
+        info: Info,
+        cohort_spec: CohortBodyInput,
+        cohort_criteria: CohortCriteriaInput | None = None,
+        dry_run: bool = False,
+    ) -> Annotated['GraphQLCohort', strawberry.lazy('api.graphql.schema')]:
+        """
+        Create a cohort with the given name and sample/sequencing group IDs.
+        """
+        from api.graphql.schema import GraphQLCohort
+
+        connection: Connection = info.context['connection']
+        connection.check_access(
```

Comment:
This check relies on the project being set on the connection, I think it will error out here